File: Algorithms/Array/Subsets.cpp
/**
 * Given a set of distinct integers, nums, return all possible subsets (the power set).

Note: The solution set must not contain duplicate subsets.

Example:

Input: nums = [1,2,3]
Output:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]
 *
 */

#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    vector<vector<int> > subsets(vector<int>& nums) {
        vector<vector<int> > result;
        result.push_back(vector<int>());
        for (int n : nums) {
            int size = result.size();
            for (int i = 0; i < size; ++i) {
                vector<int> subset = vector<int>(result[i]);
                subset.push_back(n);
                result.push_back(subset);
            }
        }

        return result;
    }
};


int main() {
    vector<int> nums = {1, 2, 3};
    Solution s;
    vector<vector<int> > ret = s.subsets(nums);

    for (int i = 0; i < ret.size(); ++i) {
        cout << "[";
        for (int j = 0; j < ret[i].size(); j++) {
            cout << ret[i][j] << ", ";
        }
        cout << "]" << endl;
    }
    cout << endl;

    return 0;
}

File: Algorithms/Array/Subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def dfs(res, nums):
            ans.append(res)
            for i in xrange(len(nums)):
                dfs(res + [nums[i]], nums[i+1:])

        ans = []
        dfs([], nums)
        return ans


if __name__ == '__main__':
    nums = [1,2,3]
    s = Solution()
    ret = s.subsets(nums)
    print ret
